mono/metadata/icall.c

```diff
@@ -770,12 +795,17 @@ ves_icall_System_Array_GetLowerBound (MonoArray *arr, gint32 dimension)
 }
 
 ICALL_EXPORT void
-ves_icall_System_Array_ClearInternal (MonoArray *arr, int idx, int length)
+ves_icall_System_Array_ClearInternal (MonoArrayHandle arr, gint32 idx, gint32 length, MonoError* error)
 {
-	int sz = mono_array_element_size (mono_object_class (arr));
-	mono_gc_bzero_atomic (mono_array_addr_with_size_fast (arr, sz, idx), length * sz);
-}
+	if (icall_array_verbose)
+		g_print ("%s arr:%p idx:%d len:%d\n", __func__, arr, (int)idx, (int)length);
 
+	guint gchandle = 0;
+	int sz = mono_array_element_size (mono_handle_class (arr));
+	gpointer p = mono_array_handle_pin_with_size (arr, sz, idx, &gchandle);
```

Comment:
It makes no sense to create a GC handle for a one shot operation that has no safepoints.

Reply:
1. How do you fixed a system.array? I couldn't figure it out. For the generic case, maybe T[] foo = this as T[] and then fixed that? That would handle some. Is the "as" cheap? What about the non-generic cases? They are always object[]? I don't know, like int[]?

2. This stuff appears to only be reachable via reflection -- see Aleksey's comments and my test code. JIT inlines it normally, so perf no matter?

Reply:
We should build higher-level operations over coop handles to encapsulate reasoning like "this is a one shot operation that has no safepoints".  Instead of reasoning at every `mono_gc_bzero_atomic (MONO_HANDLE_RAW (arr), sz)` whether it is safe or not, we should just create a function that gives it a veneer of safety.

Reply:
Yes MONO_HANDLE_RAW will need additional handling later.